doc/man3/bytes_by_prefix.3
.\"
.\" $Id: bytes_by_prefix.3,v 1.1 2000/07/27 16:59:03 alaffin Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH BYTES_BY_PREFIX 3 07/25/2000 "Linux Test Project"
.SH NAME
bytes_by_prefix \- convert a string to a int byte count
lbytes_by_prefix \- convert a string to a long byte count
llbytes_by_prefix \- convert a string to a long long byte count
.SH SYNOPSIS
int bytes_by_prefix(char *\fIstr\fR);
.br
long lbytes_by_prefix(char *\fIstr\fR);
.br
long long llbytes_by_prefix(char *\fIstr\fR);
.SH DESCRIPTION
\fBbytes_by_prefix()\fR, \fBlbytes_by_prefix()\fR, and \fBllbytes_by_prefix()\fR converts
\fIstr\fR to an integer, long, or long long  byte count.  \fIstr\fR is an
floating point number optionally followed by a single character multiplier.
Currently the following multipliers are supported:
.sp
.nf
    Char  Meaning       Multiplier
    ----  ---------     --------------------------------
    b     Blocks        BSIZE or UBSIZE
    k     Kilobytes     2^10 (1024)
    K     Kilowords     2^10 (1024) * sizeof(long)
    m     Megabytes     2^20 (1048576)
    M     Megawords     2^20 (1048576) * sizeof(long)
    g     Gigabytes     2^30 (1073741824)
    G     Gigawords     2^30 (1073741824) * sizeof(long)
.fi
.sp
\fIstr\fR is interpreted as floating point number (base 10).
When using \fBllbytes_by_prefix()\fR, the uppercase suffix will result
in multiplying by the size of a (long long) or 8.
.SH RETURNS
-1 if the integer portion of \fIstr\fR is invalid, if an unsupported
multiplier is supplied, or if \fIstr\fR has extra leading or trailing
characters.  If \fIstr\fR contains a negative number, the return
value will be negative.
.SH EXAMPLES
\fBbytes_by_prefix("1000")\fR
.br
.RS 8
Returns 1000
.RE
.br
\fBbytes_by_prefix("5b")\fR
.br
.RS 8
Returns 5 * BSIZE.
.RE
\fBbytes_by_prefix("1.5m")\fR
.br
.RS 8
Returns 1.5 * 1048576 or 1572864
.RE

.SH LIMITATIONS

\fBbytes_by_prefix()\fR and \fBlbytes_by_prefix()\fR when compiled as
a 32 bit IRIX binary can only return a max number of 2g (2147483647).
However, \fBlbytes_by_prefix()\fR is not limited by the 2g limit when
compiled as 64 bit binary, where \fBbytes_by_prefix()\fR still is limited.

Note that the size of long will vary depending how if compiled as
a 32 or 64 bit binary.  The size of a long long is always 8.

Also note that on a traditional CRAY system, a block is 4096 bytes, where
on most IRIX systems a block is 1024 bytes.



ChangeLog
LTP-20100831

commit dd59d49d75598d163f1b4599067a9f0dcb70e6c4
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 23:09:49 2010 -0700

    The cacheflush system call does not return EINVAL in the latest Linux
    kernel. see the link:
    A related patch about the cacheflush function:
    http://lkml.org/lkml/2009/4/9/203
    But Ralf had refused this patch:
    http://www.spinics.net/lists/linux-man/msg00906.html
    cacheflush01 testcase checks if the return value is EINVAL when the
    cache argument is not one of ICACHE, BCACHE or DCACHE. So it will fail
    in all boards.
    
    In this modification, checking return value SUCCESS will be added,
    instead of checking EINVAL
    
    Signed-off-by: Feng Gao <feng.gao@windriver.com>

:100644 100644 cce0b70... d0a4b1f... M	testcases/kernel/syscalls/cacheflush/cacheflush01.c

commit ebea45b92a1e30e4480f25291ef4c1ef2eeaa4bf
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 17:14:21 2010 -0700

    Get rid of unnecessary style divergence.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 0ebeca9... b08a2d9... M	testcases/commands/at/at_allow01
:100755 100755 6bca177... 9f9f433... M	testcases/commands/at/at_deny01

commit 08640ca184c1fc1e24e966c3d806a722f3ca2154
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:53:18 2010 -0700

    Add more assertions.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:000000 100644 0000000... 2c590b1... A	testcases/open_posix_testsuite/conformance/interfaces/access/assertions.xml

commit 3cdde2f86af90b3fd4e2b1468a12adfd2a9523f0
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:52:52 2010 -0700

    Note which timer failed in output.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 ac043b0... 8e149ea... M	testcases/open_posix_testsuite/conformance/behavior/timers/1-1.c

commit 207f47c4c9b4783c1a18be32431f5a1b319b9e5a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 19:34:34 2010 -0700

    Update some assertions for raise(2) to SUSv4 reqs.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 38d9542... aa47b7c... M	testcases/open_posix_testsuite/conformance/interfaces/raise/assertions.xml

commit 9c6d2b4cf4f7c166813ddee280ab2bf128eb521c
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:36:51 2010 -0700

    Fix bad syntax that I thought I hadn't committed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 f731d6c... 0ebeca9... M	testcases/commands/at/at_allow01

commit 414747ced26d943a22a3f767deb7f5b4658adb6e
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:32:18 2010 -0700

    Forgot the sleeps.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 9b10030... 6d5eb7f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 a3be617... 60e5241... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c
:100644 100644 34e3afd... abc3279... M	testcases/open_posix_testsuite/conformance/interfaces/pthread_attr_setstack/2-1.c
:100755 100755 01984c5... 536d407... M	testcases/open_posix_testsuite/scripts/generate-makefiles.sh

commit 3222fbecbfb2924a6b901a8e2048c70e1a993018
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:22:07 2010 -0700

    Remove bogus comment.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 c7d81b5... 72a1fcd... M	testcases/open_posix_testsuite/conformance/behavior/timers/2-1.c

commit 7187f0921382bf375cd111b8a480a74ce5786945
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:18:23 2010 -0700

    Repeat robustness work done to 1-1 for 1-2 and 1-3.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 a9d104c... 9b10030... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 af23ebe... a3be617... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c

commit 6c2750a55eaa30fe708c78087688c0d468bfb807
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:12:07 2010 -0700

    Make test more robust.
    
    1. Catch fork() == -1 errors.
    2. Sleep so wait(2) should always succeed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 1efd520... f46ee8f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-1.c

commit 873f6e5b28a6a23367a7117aad436378e56ad09a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:42:53 2010 -0700

    Fix ltp-aiodio.sh script so it execs aio-stress.
    
    Singed-off-by: Jacky Malcles <jacky.malcles@bull.net>

:100755 100755 ed2bce4... ab54c54... M	testscripts/ltp-aiodio.sh

commit 4fd92183833eed128bb6420ece7a6dd45dcd8ec5
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:40:51 2010 -0700

    I've prepared a patch for at_allow01 that essentially duplicates the
    fixes that were checked in for at_deny01. This fixes the script to
    actually run the test when the at utility in stead of returning TCONF,
    correcting test environment, and let the script run normally without
    changing the password. See commit
    a589d12b0b49287d10f43b1a1fb72950819588eb for more details.
    
    Signed-off-by: Henry Yei <hyei@mvista.com>

:100755 100755 cf5b97c... f731d6c... M	testcases/commands/at/at_allow01

commit 217d9b778d275bcbaf5a9fa9e1a5ac2716521cad
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:39:12 2010 -0700

    Chain cmds together to avoid intermediate errors.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 b94bd8f... 6bca177... M	testcases/commands/at/at_deny01

commit 9eeea7939bd73b295ae376dc045e5fbb984d0f38
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Wed Aug 25 20:40:23 2010 -0700

    Fix error message (was hardcoded to SIGALARM).
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>


lib/bytes_by_prefix.c
/*
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/NoticeExplan/
 */
#include <stdio.h>
#include <sys/param.h>

#include "bytes_by_prefix.h"

/****************************************************************************
 * bytes_by_prefix(s)
 *
 * Computes the number of bytes described by string s.  s is assumed to be
 * a base 10 positive (ie. >= 0) number followed by an optional single
 * character multiplier.  The following multipliers are supported:
 *
 *              char    mult
 *              -----------------
 *              b       BSIZE  or BBSIZE
 *              k       1024 bytes
 *              K       1024 * sizeof(long)
 *              m       2^20 (1048576)
 *              M       2^20 (1048576 * sizeof(long)
 *              g       2^30 (1073741824)
 *              G       2^30 (1073741824) * sizeof(long)
 *
 * for instance, "1k" and "1024" would both cause bytes_by_prefix to return 1024
 *
 * Returns -1 if mult is an invalid character, or if the integer portion of
 * s is not a positive integer.
 *
 ****************************************************************************/

#if CRAY
#define B_MULT	BSIZE		/* block size */
#elif sgi
#define B_MULT	BBSIZE		/* block size */
#elif defined(__linux__) || defined(__sun) || defined(__hpux)
#define B_MULT	DEV_BSIZE	/* block size */
#elif defined(_AIX)
#define B_MULT UBSIZE
#endif

#define K_MULT	1024		/* Kilo or 2^10 */
#define M_MULT	1048576		/* Mega or 2^20 */
#define G_MULT	1073741824	/* Giga or 2^30 */
#define T_MULT	1099511627776	/* tera or 2^40 */

int bytes_by_prefix(char *s)
{
	char mult, junk;
	int nconv;
	float num;
	int result;

	nconv = sscanf(s, "%f%c%c", &num, &mult, &junk);
	if (nconv == 0 || nconv == 3)
		return -1;

	if (nconv == 1) {
		result = num;
		return result < 0 ? -1 : result;
	}

	switch (mult) {
	case 'b':
		result = (int)(num * (float)B_MULT);
		break;
	case 'k':
		result = (int)(num * (float)K_MULT);
		break;
	case 'K':
		result = (int)((num * (float)K_MULT) * sizeof(long));
		break;
	case 'm':
		result = (int)(num * (float)M_MULT);
		break;
	case 'M':
		result = (int)((num * (float)M_MULT) * sizeof(long));
		break;
	case 'g':
		result = (int)(num * (float)G_MULT);
		break;
	case 'G':
		result = (int)((num * (float)G_MULT) * sizeof(long));
		break;
	default:
		return -1;
	}

	if (result < 0)
		return -1;

	return result;
}

long lbytes_by_prefix(char *s)
{
	char mult, junk;
	int nconv;
	float num;
	long result;

	nconv = sscanf(s, "%f%c%c", &num, &mult, &junk);
	if (nconv == 0 || nconv == 3)
		return -1;

	if (nconv == 1) {
		result = (long)num;
		return result < 0 ? -1 : result;
	}

	switch (mult) {
	case 'b':
		result = (long)(num * (float)B_MULT);
		break;
	case 'k':
		result = (long)(num * (float)K_MULT);
		break;
	case 'K':
		result = (long)((num * (float)K_MULT) * sizeof(long));
		break;
	case 'm':
		result = (long)(num * (float)M_MULT);
		break;
	case 'M':
		result = (long)((num * (float)M_MULT) * sizeof(long));
		break;
	case 'g':
		result = (long)(num * (float)G_MULT);
		break;
	case 'G':
		result = (long)((num * (float)G_MULT) * sizeof(long));
		break;
	default:
		return -1;
	}

	if (result < 0)
		return -1;

	return result;
}

/*
 * Force 64 bits number when compiled as 32 IRIX binary.
 * This allows for a number bigger than 2G.
 */
long long llbytes_by_prefix(char *s)
{
	char mult, junk;
	int nconv;
	double num;
	long long result;

	nconv = sscanf(s, "%lf%c%c", &num, &mult, &junk);
	if (nconv == 0 || nconv == 3)
		return -1;
	if (nconv == 1) {
		result = (long long)num;
		return result < 0 ? -1 : result;
	}

	switch (mult) {
	case 'b':
		result = (long long)(num * (float)B_MULT);
		break;
	case 'k':
		result = (long long)(num * (float)K_MULT);
		break;
	case 'K':
		result = (long long)((num * (float)K_MULT) * sizeof(long long));
		break;
	case 'm':
		result = (long long)(num * (float)M_MULT);
		break;
	case 'M':
		result = (long long)((num * (float)M_MULT) * sizeof(long long));
		break;
	case 'g':
		result = (long long)(num * (float)G_MULT);
		break;
	case 'G':
		result = (long long)((num * (float)G_MULT) * sizeof(long long));
		break;
	default:
		return -1;
	}

	if (result < 0)
		return -1;

	return result;
}
